[mlir][spirv] Fix some issues related to converting ub.poison to SPIR-V

andfau-amd opened the pull request:
This is a follow-up to 5df62bdc9be9c258c5ac45c8093b71e23777fa0e. That commit should not have needed to make the vector.insert and vector.extract conversions to SPIR-V directly handle the static poison index case, as there is a fold from those to ub.poison, and a conversion pattern from ub.poison to spirv.Undef, however:

- The ub.poison fold result could not be materialized by the vector dialect (fixed as of d13940ee263ff50b7a71e21424913cc0266bf9d4).
- The conversion pattern wasn't being populated in VectorToSPIRVPass, which is used by the tests. This commit changes this.
- The ub.poison to spirv.Undef pattern rejected non-scalar types, which prevented its use for vector results. It is unclear why this restriction existed; a remark in D156163 said this was to avoid converting "user types", but it is not obvious why these shouldn't be permitted (the SPIR-V specification allows OpUndef for all types except OpTypeVoid). This commit removes this restriction.

With these fixed, this commit removes the redundant static poison index handling, and updates the tests.

andfau-amd commented on mlir/test/lib/Pass/TestConvertToSPIRVPass.cpp:
```diff
@@ -93,6 +94,7 @@ struct TestConvertToSPIRVPass final
   }
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<spirv::SPIRVDialect>();
+    registry.insert<ub::UBDialect>();
```
Oops, my commit message doesn't match my actual patch, I forgot that I didn't actually have to add the UB conversion patterns here, but I _did_ have to add a dialect here to prevent it from crashing.

andfau-amd commented on mlir/test/Conversion/VectorToSPIRV/vector-to-spirv.mlir:
```diff
@@ -175,9 +175,11 @@ func.func @extract(%arg0 : vector<2xf32>) -> (vector<1xf32>, f32) {
 
+// CHECK-LABEL: @extract_poison_idx
+//  CHECK-SAME: %[[ARG0:.+]]: vector<4xf32>
+//       CHECK:   %[[R:.+]] = vector.extract %[[ARG0]][-1] : f32 from vector<4xf32>
```
I'm assuming here that `-vector-to-spirv` should not be trying to use the UBToSPIRV patterns, but I'm not sure if that's reasonable on my part.

kuhar replied:
We can add poison conversion to vector-to-spirv -- it's not a big deal unless it break the layering / linking. If it does, we can create a test pass that runs both vector and ub conversions

kuhar commented on mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:
```diff
@@ -182,17 +182,12 @@ struct VectorExtractOpConvert final
-            rewriter.getI32ArrayAttr(id.value()));
+      // Static use of the poison index is handled elsewhere (folded to poison).
+      if (id == vector::ExtractOp::kPoisonIndex)
+        return failure();
```
Maybe move the comment above to `return rewriter.notifyMatchFailure("reason");` so that this shows up in debug logs?

andfau-amd replied:
It doesn't show up in `-debug-only=pattern-application` or `-debug-only=dialect-conversion` for example.

andfau-amd replied:
Ooh, it does show up for the latter if I remove the UB-to-SPIR-V pattern again. Great!

kuhar reviewed (approved):
LGTM % one suggestion

andfau-amd committed:
[mlir][spirv] Fix some issues related to converting ub.poison to SPIR-V

This is a follow-up to 5df62bdc9be9c258c5ac45c8093b71e23777fa0e. That
commit should not have needed to make the vector.insert and
vector.extract conversions to SPIR-V directly handle the static poison
index case, as there is a fold from those to ub.poison, and a conversion
pattern from ub.poison to spirv.Undef, however:

- The ub.poison fold result could not be materialized by the vector
  dialect (fixed as of d13940ee263ff50b7a71e21424913cc0266bf9d4).
- The conversion pattern wasn't being populated in VectorToSPIRVPass,
  which is used by the tests. This commit changes this.
- The ub.poison to spirv.Undef pattern rejected non-scalar types, which
  prevented its use for vector results. It is unclear why this
  restriction existed; a remark in D156163 said this was to avoid
  converting "user types", but it is not obvious why these shouldn't
  be permitted (the SPIR-V specification allows OpUndef for all types
  except OpTypeVoid). This commit removes this restriction.

With these fixed, this commit removes the redundant static poison index
handling, and updates the tests.

Hardcode84 reviewed (approved):
> @Hardcode84 I added you as a reviewer here because you seem to be the one that added this restriction against converting ub.poison to spirv.Undef for some types.

Yeah, I'm fine with removing the restriction

andfau-amd commented:
Thanks!
